### PR TITLE
Fix pickling for provider models by removing SimpleNamespace

### LINKED ISSUE #4550: Cannot copy a pulse enabled backend configuration
### Information

- **Qiskit Terra version**: master
- **Python version**:
- **Operating system**:

### What is the current behavior?
Doing `copy.deepcopy(provider.backends.ibmq_almaden.configuration())` fails with the error:

TypeError: __init__() missing 10 required positional arguments: 'n_uchannels', 'u_channel_lo', 'meas_levels', 'qubit_lo_range', 'meas_lo_range', 'dt', 'dtm', 'rep_times', 'meas_kernels', and 'discriminators'

Non-pulse enabled backends work fine. e.g.

```python
copy.deepcopy(provider.backends.ibmq_vigo.configuration())
```

QasmBackendConfiguration(backend_name='ibmq_vigo', backend_version='1.0.2', basis_gates=['u1', 'u2', 'u3', 'cx', 'id'], conditional=False, coupling_map=[[0, 1], [1, 0], [1, 2], [1, 3], [2, 1], [3, 1], [3, 4], [4, 3]], gates=[GateConfig(id, [], gate id q { U(0,0,0) q; }, [[0], [1], [2], [3], [4]]), GateConfig(u1, ['lambda'], gate u1(lambda) q { U(0,0,lambda) q; }, [[0], [1], [2], [3], [4]]), GateConfig(u2, ['phi', 'lambda'], gate u2(phi,lambda) q { U(pi/2,phi,lambda) q; }, [[0], [1], [2], [3], [4]]), GateConfig(u3, ['theta', 'phi', 'lambda'], gate u3(theta,phi,lambda) q { U(theta,phi,lambda) q; }, [[0], [1], [2], [3], [4]]), GateConfig(cx, [], gate cx q1,q2 { CX q1,q2; }, [[0, 1], [1, 0], [1, 2], [1, 3], [2, 1], [3, 1], [3, 4], [4, 3]])], local=False, max_shots=8192, memory=True, n_qubits=5, open_pulse=False, simulator=False)

### PR DESCRIPTION
**Pull request by mtreinish**
### Summary

The provider model classes were built as subclasses for
`SimpleNamespace`, primarily because they allow attribute access for
arbitrary kwargs from the constructor and `SimpleNamespace` makes that
easy. Unfortunately this adds a great deal of complexity if the subclass
has required positional arguments, since the parent class signature for
`SimpleNamespace` is just **kwargs. For example, when using pickle to
serialize the object (often as part of a deecopy or multiprocessing) it
relies on the `__reduce__` from the parent class which is configured to
only handle **kwargs and it will fail because of missing required
positional arguments. `QasmBackendConfiguration `attempted to fix this by
defining a custom `__reduce__` method to pass in the required args, but
then pickle would miss the kwargs resulting in only a partial copy. This
wasn't an issue in practice because there were not too many extra kwargs
being passed in practice (although there were some edge cases caused by
this). Where this came into play more was for the `PulseBackendConfiguration`
class. It  was not defining it's own `__reduce__` method so it was relying on
the definition in it's parent class `QasmBackendConfiguration`. However,
because `PulseBackendConfiguration` has additional required position arguments
on top of what `QasmBackendConfiguration` requires pickle was not able to
create a new object when loading. When adding a custom `__reduce__` method
with the extra required args for pulse it appears to work, but the loss
of arbitrary kwargs causes an issue in practice since they're heavily
used in pulse.

This commit fixes this by removing the `SiimpleNamespace` usage and making
the provider model classes just a bare object. The kwarg based
attribute flow is built manually by storing everything in a private dict and
having a custom `_getattr__` method to handle attribute access. This
maintains compatibility with the expected access patterns, but fixes the
deepcopy issue.

### Details and comments

Fixes #4550

### CODE REVIEW

File: qiskit/providers/models/backendproperties.py [hunk 1]
```diff
@@ -272,6 +271,7 @@ def to_dict(self):
                            'last_update_date', 'qubits', 'general', 'gates',
                            '_gates', '_qubits']:
                 out_dict[key] = value
+        out_dict.update(self._data)
```
Comment (jyu00) [inline qiskit/providers/models/backendproperties.py:273]: Do we still need the `for` loop at line 269-273?

Comment (mtreinish) [reply qiskit/providers/models/backendproperties.py:273]: Nope good call. I'll fix it now

Comment (mtreinish) [reply qiskit/providers/models/backendproperties.py:268]: Fixed in: aabd88aa

File: qiskit/providers/models/pulsedefaults.py [hunk 2]
```diff
@@ -133,6 +143,7 @@ def to_dict(self):
         for key, value in self.__dict__.items():
             if key not in ['name', 'qubits', 'sequence']:
                 out_dict[key] = value
+        out_dict.update(self._data)
```
[the remark this answers is not in the document]
Comment (mtreinish) [reply qiskit/providers/models/pulsedefaults.py:141]: Fixed in: aabd88aa

**Pull request merged**